File: Systems-for-Design-and-Implementation/Labs/Lab9/ProiectCSharp/TransportPersistence/transport.persistence/repository/SeatDBRepository.cs
using System.Collections.Generic;
using System.Data;
using log4net;
using Transport.Model.transport.model;
using TransportModel.transport.model;

namespace TransportPersistence.transport.persistence
{
    public class SeatDBRepository : ISeatRepository
    {
        private static readonly ILog logger = LogManager.GetLogger(typeof(SeatDBRepository));
        private ReservationDBRepository reservationRepository;

        public SeatDBRepository()
        {
            logger.Info("Initializing SeatDBRepository");
            reservationRepository = new ReservationDBRepository();
        }

        public Seat FindOne(long id)
        {
            logger.Info("Finding Seat with id " + id);
            IDbConnection connection = DbUtils.getConnection();
            using (IDbCommand command = connection.CreateCommand())
            {
                command.CommandText = "SELECT * FROM Seats WHERE id = @id";
                IDataParameter idParam = command.CreateParameter();
                idParam.ParameterName = "@id";
                idParam.Value = id;
                command.Parameters.Add(idParam);
                using (IDataReader result = command.ExecuteReader())
                {
                    if (result.Read())
                    {
                        // RESERVATION COLUMN
                        long reservationId = result.GetInt64(1);
                        Reservation reservation = reservationRepository.FindOne(reservationId);
                        if (reservation == null)
                            return null;
                        
                        long seatNumber = result.GetInt64(2);
                        Seat seat = new Seat(reservation, seatNumber);
                        seat.id = id;
                        return seat;
                    }
                }
            }
            return null;
        }

        public IEnumerable<Seat> FindAll()
        {
            logger.Info("Finding all Seats");
            IDbConnection connection = DbUtils.getConnection();
            List<Seat> seats = new List<Seat>();
            using (IDbCommand command = connection.CreateCommand())
            {
                command.CommandText = "SELECT * FROM Seats";
                using (IDataReader result = command.ExecuteReader())
                {
                    while (result.Read())
                    {
                        long id = result.GetInt64(0);
                        Seat seat = FindOne(id);
                        if (seat != null)
                            seats.Add(seat);
                    }
                }
            }
            return seats;
        }
        
        public IEnumerable<Seat> FindSeatsByReservationId(long reservationId)
        {
            logger.Info("Finding Seats by reservation id " + reservationId);
            IDbConnection connection = DbUtils.getConnection();
            List<Seat> seats = new List<Seat>();
            using (IDbCommand command = connection.CreateCommand())
            {
                command.CommandText = "SELECT * FROM Seats WHERE reservation_id = @reservationId";
                IDataParameter reservationIdParam = command.CreateParameter();
                reservationIdParam.ParameterName = "@reservationId";
                reservationIdParam.Value = reservationId;
                command.Parameters.Add(reservationIdParam);
                using (IDataReader result = command.ExecuteReader())
                {
                    while (result.Read())
                    {
                        Reservation reservation = reservationRepository.FindOne(reservationId);
                        if (reservation == null)
                            return null;

                        long id = result.GetInt64(0);
                        long seatNumber = result.GetInt64(2);
                        Seat seat = new Seat(reservation, seatNumber);
                        seat.id = id;
                        seats.Add(seat);
                    }
                }
            }
            return seats;
        }

        public List<Seat> FindSeatsByTripId(long tripId)
        {
            logger.Info("Finding Seats by trip id " + tripId);
            IDbConnection connection = DbUtils.getConnection();
            List<Seat> seats = new List<Seat>();
            using (IDbCommand command = connection.CreateCommand())
            {
                command.CommandText = "SELECT Seats.* FROM Seats INNER JOIN Reservations ON Seats.reservation_id = Reservations.id WHERE Reservations.trip_id = @tripId";
                IDataParameter tripIdParam = command.CreateParameter();
                tripIdParam.ParameterName = "@tripId";
                tripIdParam.Value = tripId;
                command.Parameters.Add(tripIdParam);
                using (IDataReader result = command.ExecuteReader())
                {
                    while (result.Read())
                    {
                        long reservationId = result.GetInt64(1);
                        Reservation reservation = reservationRepository.FindOne(reservationId);
                        if (reservation == null)
                            return null;

                        long id = result.GetInt64(0);
                        long seatNumber = result.GetInt64(2);
                        Seat seat = new Seat(reservation, seatNumber);
                        seat.id = id;
                        seats.Add(seat);
                    }
                }
            }
            return seats;
        }

        public Seat Save(Seat seat)
        {
            logger.Info("Not Implemented!");
            return null;
        }

        public Seat Delete(long id)
        {
            logger.Info("Not Implemented!");
            return null;
        }

        public Seat Update(Seat seat)
        {
            logger.Info("Not Implemented!");
            return null;
        }
    }
}

File: Systems-for-Design-and-Implementation/Labs/Lab7/ProiectJava/TransportPersistence/src/main/java/transport/persistence/repository/SeatDBRepository.java
package transport.persistence.repository;

import transport.model.Reservation;
import transport.model.Seat;
import transport.persistence.ISeatRepository;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.Properties;

public class SeatDBRepository implements ISeatRepository {

    private JdbcUtils dbUtils;
    private ReservationDBRepository reservationRepository;
    //private static final Logger logger = LogManager.getLogger();

    public SeatDBRepository(Properties props) {
        //logger.info("Initializing SeatDBRepository with properties: {}", props);
        dbUtils = new JdbcUtils(props);
        reservationRepository = new ReservationDBRepository(props);
    }

    @Override
    public Optional<Seat> findOne(Long id) {
        //logger.traceEntry();
        Connection con = dbUtils.getConnection();
        try (PreparedStatement preStmt = con.prepareStatement("select * from Seats where id=?")) {
            preStmt.setLong(1, id);
            try (ResultSet result = preStmt.executeQuery()) {
                if (result.next()) {
                    Long reservationId = result.getLong("reservation_id");
                    Reservation reservation = reservationRepository.findOne(reservationId).orElse(null);
                    if (reservation == null) {
                        return Optional.empty();
                    }
                    Long seatNumber = result.getLong("seat_number");

                    Seat seat = new Seat(reservation, seatNumber);
                    seat.setId(id);
                    return Optional.of(seat);
                }
            }
        } catch (SQLException e) {
            //logger.error(e);
            System.out.println("Error DB " + e);
        }
        return Optional.empty();
    }

    @Override
    public Iterable<Seat> findAll() {
        //logger.traceEntry();
        Connection con = dbUtils.getConnection();
        List<Seat> seats = new ArrayList<>();
        try (PreparedStatement preStmt = con.prepareStatement("select * from Seats")) {
            try (ResultSet result = preStmt.executeQuery()) {
                while (result.next()) {
                    Long id = result.getLong("id");
                    Seat seat = findOne(id).orElse(null);
                    if (seat != null) {
                        seats.add(seat);
                    }
                }
            }
        } catch (SQLException e) {
            //logger.error(e);
            System.out.println("Error DB " + e);
        }
        return seats;
    }

    @Override
    public Iterable<Seat> findSeatsByReservationId(long reservationId) {
        //logger.traceEntry();
        Connection con = dbUtils.getConnection();
        List<Seat> seats = new ArrayList<>();
        try (PreparedStatement preStmt = con.prepareStatement("select * from Seats where reservation_id=?")) {
            preStmt.setLong(1, reservationId);
            try (ResultSet result = preStmt.executeQuery()) {
                while (result.next()) {
                    Reservation reservation = reservationRepository.findOne(reservationId).orElse(null);
                    Long id = result.getLong("id");
                    Long seatNumber = result.getLong("seat_number");
                    Seat seat = new Seat(reservation, seatNumber);
                    seat.setId(id);
                    if (seat != null) {
                        seats.add(seat);
                    }
                }
            }
        } catch (SQLException e) {
            //logger.error(e);
            System.out.println("Error DB " + e);
        }
        return seats;
    }

    @Override
    public List<Seat> findSeatsByTripId(Long tripId) {
        //logger.traceEntry();
        Connection con = dbUtils.getConnection();
        List<Seat> seats = new ArrayList<>();
        try (PreparedStatement preStmt = con.prepareStatement("select Seats.* from Seats " +
                "inner join Reservations on Seats.reservation_id = Reservations.id " +
                "where Reservations.trip_id = ?")) {
            preStmt.setLong(1, tripId);
            try (ResultSet result = preStmt.executeQuery()) {
                while (result.next()) {
                    Long reservationId = result.getLong("reservation_id");
                    Reservation reservation = reservationRepository.findOne(reservationId).orElse(null);
                    if (reservation == null) {
                        return null;
                    }
                    Long id = result.getLong("id");
                    Long seatNumber = result.getLong("seat_number");
                    Seat seat = new Seat(reservation, seatNumber);
                    seat.setId(id);
                    seats.add(seat);
                }
            }
        } catch (SQLException e) {
            //logger.error(e);
            System.out.println("Error DB " + e);
        }
        return seats;
    }

    @Override
    public Optional<Seat> save(Seat seat) {
        return Optional.empty();
    }

    @Override
    public Optional<Seat> delete(Long id) {
        return Optional.empty();
    }

    @Override
    public Optional<Seat> update(Seat seat) {
        return Optional.empty();
    }
}